packages/flutter/lib/src/widgets/restoration.dart

```diff
@@ -0,0 +1,1031 @@
+///           children: <Widget>[
+///             Text(
+///               'You have pushed the button this many times:',
+///             ),
+///             Text(
+///               '${_counter.value}',
+///               style: Theme.of(context).textTheme.headline4,
+///             ),
+///           ],
+///         ),
+///       ),
+///       floatingActionButton: FloatingActionButton(
+///         onPressed: _incrementCounter,
+///         tooltip: 'Increment',
+///         child: Icon(Icons.add),
+///       ),
+///     );
+///   }
+/// }
+/// ```
+/// {@end-tool}
+///
+/// See also:
+///
+///  * [RestorableProperty], which is the base class for all restoration
+///    properties managed by this mixin.
+///  * [RestorationManager], which describes how state restoration in Flutter
+///    works.
+///  * [RestorationScope], which creates a new namespace for [RestorationId]s
+///    in the widget tree.
+@optionalTypeArgs
+mixin RestorationMixin<S extends StatefulWidget> on State<S> {
+  /// The [RestorationId] used for the [RestorationBucket] in which the mixin
+  /// will store the restoration data of all registered properties.
+  ///
+  /// The restoration ID is used to claim a child [RestorationScope] from the
+  /// surrounding [RestorationScope] (accessed via [RestorationScope.of]) and
+  /// the ID must be unique in that scope (otherwise an exception is triggered
+  /// in debug move).
+  ///
+  /// State restoration for this mixin is turned off when this getter returns
+  /// null or when there is no surrounding [RestorationScope] available. When
+  /// state restoration is turned off, the values of the registered properties
+  /// cannot be restored.
+  ///
+  /// Whenever the value returned by this getter changes,
+  /// [didUpdateRestorationId] must be called unless the (unless the change
+  /// already triggered a call to [didUpdateWidget]).
+  ///
+  /// The restoration ID returned by this getter is often provided in the
+  /// constructor of the [StatefulWidget] that this [State] object is associated
+  /// with.
+  @protected
+  RestorationId get restorationId;
+
+  /// The [RestorationBucket] used for the restoration data of the
+  /// [RestorableProperty]s registered to this mixin.
+  ///
+  /// The bucket has been claimed from the surrounding [RestorationScope] using
+  /// [restorationId].
+  ///
+  /// The getter returns null if state restoration is turned off. When state
+  /// restoration is turned on or off during the lifetime of this mixin (and
+  /// hence the return value of this getter switches between null and non-null)
+  /// [didToggleBucket] is called.
+  ///
+  /// Interacting directly with this bucket is uncommon. However, the bucket may
+  /// be injected into the widget tree in the [State]'s `build` method using a
+  /// [UnmanagedRestorationScope]. That allows descendants to claim child
+  /// buckets from this bucket for their own restoration needs.
```

Comment:
this would be a good place for sample code (showing the use of UnmanagedRestorationScope)

Reply:
Made a note. I am going to add more samples in a follow-up since this PR is already gigantic.